File: initPPG_struct.m
function [ PPG_struct ] = initPPG_struct()

% ------------------ KNOBS ------------------------------------    
    
    PPG_struct.Knob_Threshold_Goal      = uint8(5);         % Target threshold level
    PPG_struct.Knob_ThresholdTC         = single(0.15);     % Time const used in moving mean filter
    PPG_struct.Knob_Qualified_Threshold = uint16(7);        % Pulse amplitude qualification threshold for Pulse Rate Filter
    PPG_struct.Knob_MeanPulseRateTC     = single(0.85);     % Time Const of Mean Pulse Rate
                                                            
    PPG_struct.Knob_MinPulseRate        = uint8(20);        % for Pulse Rate qualification
    PPG_struct.Knob_LightExposureGoal   = uint8(90);
    PPG_struct.Knob_LightExposureGoal_H = uint8(99);
    PPG_struct.Knob_LightExposureGoal_L = uint8(45);

    PPG_struct.Knob_BiasUpdateInterval  = uint8(50);
    PPG_struct.Knob_AdaptiveBiasGain    = uint8(16);
    PPG_struct.Knob_MedianTaps          = uint8(5);
    
    
% ---------- PulseRate Calculations ----------------------   
    PPG_struct.PulseRate                = uint8(0);
    PPG_struct.FilteredDataInput        = single(0);
    PPG_struct.PreviousInputVal         = single(0);
    PPG_struct.Quadrant                 = uint8(1);
    PPG_struct.PreviousQTime            = zeros([4,1],'single');
    PPG_struct.LocalHigh                = single(0);
    PPG_struct.LocalLow                 = single(0);
    PPG_struct.Threshold                = single(10); 
    PPG_struct.Timestamp                = uint16(0);
    PPG_struct.PreviousTimestamp        = uint16(0);
    PPG_struct.IgnoreFirstPulses        = int8(6);
    PPG_struct.NumberOfLEDPulses        = uint8(0);
    
% ---------- PulseRate Filter Variables ----------------------

    PPG_struct.PulseRateHistory         = zeros(1,5,'single');
    PPG_struct.PulseRatePointer         = uint8(1);
    PPG_struct.MovingPRMemory           = single(0);
    PPG_struct.LastQ1PulseRate          = uint8(0);
    PPG_struct.MeanPulseRate            = single(0);

File: codegen/codegen/lib/Process_NewPPGData/initPPG_struct.c
/*
 * initPPG_struct.c
 *
 * Code generation for function 'initPPG_struct'
 *
 */

/* Include files */
#include "rt_nonfinite.h"
#include "Process_NewPPGData.h"
#include "PulseRateFilter_Experimental.h"
#include "initFilt_struct.h"
#include "initPPG_struct.h"
#include "stillness_update_eff.h"

/* Function Definitions */
struct0_T initPPG_struct(void)
{
  struct0_T PPG_struct;
  int i;

  /*   UNTITLED Summary of this function goes here */
  /*    Detailed explanation goes here */
  /*  ------------------ KNOBS ------------------------------------     */
  PPG_struct.Knob_Threshold_Goal = 150;

  /*  Target threshold level */
  PPG_struct.Knob_ThresholdTC = 0.05F;

  /*  Time const used in moving mean filter */
  PPG_struct.Knob_Qualified_Threshold = 15;

  /*  Pulse amplitude qualification threshold for Pulse Rate Filter */
  PPG_struct.Knob_MeanPulseRateTC = 0.985F;
  PPG_struct.Knob_MinPulseRate = 20;

  /*  for Pulse Rate qualification */
  PPG_struct.Knob_LightExposureGoal = 180;
  PPG_struct.Knob_LightExposureGoal_H = 200;
  PPG_struct.Knob_LightExposureGoal_L = 90;
  PPG_struct.Knob_BiasUpdateInterval = 25;
  PPG_struct.Knob_AdaptiveBiasGain = 16;

  /* PPG_struct.Knob_ThreshGoalGain      = single(7.0); */
  PPG_struct.Knob_MedianTaps = 5;

  /*  ---------- PulseRate Calculations ----------------------    */
  PPG_struct.PulseRate = 0;
  PPG_struct.FilteredDataInput = 0.0F;
  PPG_struct.PreviousInputVal = 0.0F;
  PPG_struct.Quadrant = 1;
  for (i = 0; i < 4; i++) {
    PPG_struct.PreviousQTime[i] = 0.0F;
  }

  PPG_struct.LocalHigh = 0.0F;
  PPG_struct.LocalLow = 0.0F;
  PPG_struct.Threshold = 0.0F;
  PPG_struct.Timestamp = 0;
  PPG_struct.PreviousTimestamp = 0;
  PPG_struct.IgnoreFirstPulses = 6;
  PPG_struct.NumberOfLEDPulses = 0;

  /*  ---------- PulseRate Filter Variables ---------------------- */
  for (i = 0; i < 5; i++) {
    PPG_struct.PulseRateHistory[i] = 0.0F;
  }

  PPG_struct.PulseRatePointer = 1;
  PPG_struct.MovingPRMemory = 0.0F;
  PPG_struct.LastQ1PulseRate = 0;
  PPG_struct.MeanPulseRate = 0.0F;
  return PPG_struct;
}

/* End of code generation (initPPG_struct.c) */
